In [2]:
import pandas as pd
from sqlalchemy import create_engine
import urllib.parse


In [3]:
password = urllib.parse.quote_plus("N@ndini2005")

engine = create_engine(
    f"postgresql+psycopg2://postgres:{password}@localhost:5432/ecommerce_db"
)

conn = engine.connect()
print("CONNECTED ✅")
conn.close()


CONNECTED ✅


In [7]:
df_final = pd.read_sql(
    "SELECT * FROM customer_behavior_final",
    engine
)

print("Data loaded ✅")
df_final.shape


Data loaded ✅


(1000000, 66)

In [8]:
segment_distribution = (
    df_final["customer_segment"]
    .value_counts(normalize=True) * 100
)

segment_distribution


customer_segment
general_segment    92.0162
target_segment      7.9838
Name: proportion, dtype: float64

In [9]:
df_final.groupby("customer_segment")["engagement_score"].mean()


customer_segment
general_segment    13.144588
target_segment     17.645094
Name: engagement_score, dtype: float64

In [10]:
df_final.groupby("customer_segment")["impulse_buying_score"].mean()


customer_segment
general_segment    4.853267
target_segment     6.671397
Name: impulse_buying_score, dtype: float64

In [11]:
df_final["recommended_category"].value_counts(normalize=True) * 100


recommended_category
home_appliances    54.5613
electronics        45.4387
Name: proportion, dtype: float64

In [14]:
segment_summary.to_sql(
    "final_business_insights",
    engine,
    if_exists="replace"
)

print("Business insights table saved ✅")

Business insights table saved ✅


In [13]:
segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary



,engagement_score,impulse_buying_score,return_rate,premium_subscription
customer_segment,,,,
general_segment,13.144588,4.853267,50.011078,0.359515
target_segment,17.645094,6.671397,49.934304,0.358263


In [13]:
segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary



,engagement_score,impulse_buying_score,return_rate,premium_subscription
customer_segment,,,,
general_segment,13.144588,4.853267,50.011078,0.359515
target_segment,17.645094,6.671397,49.934304,0.358263


In [13]:
segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary



,engagement_score,impulse_buying_score,return_rate,premium_subscription
customer_segment,,,,
general_segment,13.144588,4.853267,50.011078,0.359515
target_segment,17.645094,6.671397,49.934304,0.358263


In [13]:
segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary



,engagement_score,impulse_buying_score,return_rate,premium_subscription
customer_segment,,,,
general_segment,13.144588,4.853267,50.011078,0.359515
target_segment,17.645094,6.671397,49.934304,0.358263


In [13]:
segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary



,engagement_score,impulse_buying_score,return_rate,premium_subscription
customer_segment,,,,
general_segment,13.144588,4.853267,50.011078,0.359515
target_segment,17.645094,6.671397,49.934304,0.358263


In [13]:
segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary



,engagement_score,impulse_buying_score,return_rate,premium_subscription
customer_segment,,,,
general_segment,13.144588,4.853267,50.011078,0.359515
target_segment,17.645094,6.671397,49.934304,0.358263


segment_summary = df_final.groupby("customer_segment").agg({
    "engagement_score": "mean",
    "impulse_buying_score": "mean",
    "return_rate": "mean",
    "premium_subscription": "mean"
})

segment_summary


In [16]:
# Create recommendation mapping from segment to top category
segment_recommendation = (
    df_final.groupby("customer_segment")["recommended_category"]
    .agg(lambda x: x.value_counts().index[0])
    .to_dict()
)

segment_recommendation

{'general_segment': 'home_appliances', 'target_segment': 'electronics'}

In [23]:
def recommend_category(customer_segment):
    return segment_recommendation.get(customer_segment, "No Recommendation")

In [20]:
recommend_category("target_segment")

'electronics'

# 📊 Step 4: Business Insights Analysis

## 🔎 Segment Comparison

### 1️⃣ Engagement Score
- Target Segment: 17.64
- General Segment: 13.14  
✅ Target customers are more engaged.

### 2️⃣ Impulse Buying Score
- Target Segment: 6.67
- General Segment: 4.85  
✅ Target customers show higher impulse buying behavior.

### 3️⃣ Return Rate
- Both segments have similar return rates (~50%)  
⚠ No major difference in returns.

### 4️⃣ Premium Subscription Rate
- Both segments around 35%  
📌 Subscription behavior is similar.

---

## 🎯 Business Recommendation

- Promote **electronics**(Top category:54%) to Target Segment users.
- Run personalized marketing campaigns.
- Increase premium subscription benefits for high engagement users.
- Monitor return rates to reduce product returns.
- Segment-based rule engine recommending top category per segment.

---

## 📈 Conclusion

Target Segment customers are:
✔ More engaged  
✔ Higher impulse buyers  
✔ Good candidates for premium product promotions  

This segment should be the primary focus for revenue growth


In [3]:
%who

Interactive namespace is empty.


In [4]:
df_final.head()

NameError: name 'df_final' is not defined

In [5]:
import psycopg2
import pandas as pd

conn = psycopg2.connect(
    host="localhost",
    database="ecommerce_db",
    user="postgres",
    password="N@ndini2005"
)

query = "SELECT * FROM customer_behavior_cleaned LIMIT 5;"
df = pd.read_sql(query, conn)

df.head()

C:\Users\Nandini\AppData\Local\Temp\ipykernel_11392\4216624916.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


,user_id,age,gender,country,urban_rural,income_level,employment_status,education_level,relationship_status,has_children,...,app_usage_frequency,notification_response_rate,account_age_months,last_purchase_date,social_sharing_frequency,premium_subscription,return_rate,age_group,income_band,engagement_score
0,1,56,female,germany,suburban,90860,self-employed,associate degree,single,0,...,7,74,19,2025-06-22,6,1,50,51+,medium,7
1,2,69,male,japan,suburban,35423,unemployed,bachelor,single,1,...,5,23,8,2026-07-25,3,0,37,51+,low,12
2,3,46,female,india,urban,21467,self-employed,associate degree,married,1,...,7,12,13,2026-02-26,6,0,53,36-50,low,6
3,4,32,male,canada,urban,41770,self-employed,bachelor,widowed,0,...,4,19,9,2026-10-27,7,0,98,26-35,low,3
4,5,60,female,japan,urban,183882,employed,associate degree,widowed,1,...,7,30,3,2026-06-23,3,0,86,51+,high,23
